core/iwasm/common/wasm_runtime_common.c

```diff
@@ -7739,3 +7748,114 @@ wasm_runtime_is_underlying_binary_freeable(WASMModuleCommon *const module)
 
     return true;
 }
+
+/*TODO: take us(below) out when have a linker */
+void
+wasm_runtime_destroy_imports(WASMExternInstance *extern_inst)
+{
+    if (extern_inst == NULL)
+        return;
+
+    wasm_runtime_free(extern_inst);
+}
+
+WASMExternInstance *
+wasm_runtime_create_imports(WASMModuleCommon *module,
+                            bool (*module_name_filter)(const char *))
+{
+    int32 import_count = wasm_runtime_get_import_count(module);
+    WASMExternInstance *imports = NULL;
+
+    if (import_count == 0)
+        return NULL;
+
+    imports = runtime_malloc(sizeof(WASMExternInstance) * import_count,
+                             NULL, // module_inst
+                             NULL, 0);
+    if (!imports) {
+        LOG_ERROR("allocate memory failed");
+        return NULL;
+    }
+
+    for (int32 i = 0; i < import_count; i++) {
+        wasm_import_t import_type = { 0 };
+        wasm_runtime_get_import_type(module, i, &import_type);
+
+        if (module_name_filter
+            && !module_name_filter(import_type.module_name)) {
+            LOG_DEBUG("skip import(%s,%s)\n", import_type.module_name,
+                      import_type.name);
+            continue;
+        }
+
+        LOG_DEBUG("create import(%s,%s) kind %d\n", import_type.module_name,
+                  import_type.name, import_type.kind);
+        WASMExternInstance *extern_instance = imports + i;
+        extern_instance->module_name = import_type.module_name;
+        extern_instance->field_name = import_type.name;
+        extern_instance->kind = import_type.kind;
+
+        if (import_type.kind == WASM_IMPORT_EXPORT_KIND_MEMORY) {
+            extern_instance->u.memory = wasm_runtime_create_memory(
+                module, import_type.u.memory_type, 0);
```

Comment:
Should check the return value? If it is NULL, should call wasm_runtime_destroy_imports to free the memory and return NULL?